### PR TITLE
About adding image upload function

### PR DESCRIPTION
**Pull request by kazutan1230**
### What does this PR do?

I added an image upload function.
Ther is no function to submit images and FormData in the buttons.

New file
-/components/imageInput.tsx
-/components/sendImageForm.tsx
Changed file
-/components/uploadFormHook.tsx

- [x] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

By localhost server and github workflows.

### CODE REVIEW

**Comment by Marukome0743** [comment]
> ### What does this PR do?
> I added an image upload function. Ther is no function to submit images and FormData in the bottons.

Do you mean `buttons` instead of `bottons`?
Or `bottom`?

File: components/imageInput.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import type React from 'react'
+import type { InputHTMLAttributes } from 'react'
+// import type { RefCallBack } from 'react-hook-form'
```
Comment (Marukome0743) [inline components/imageInput.tsx:3]: Remove this comment out.

File: components/imageInput.tsx [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+export type Props = {
```
Comment (Marukome0743) [inline components/imageInput.tsx:5]: This Props is not exported.
Then remove `export`
Just `type Props`

File: components/sendImageForm.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
+'use client'
+
+import type React from 'react'
+import { useRef, useState } from 'react'
+import { ImageInput } from './imageInput'
+
+const IMAGE_ID = 'image'
```
Comment (Marukome0743) [inline components/sendImageForm.tsx:7]:
```suggestion
// TODO: create and assign unique id
const IMAGE_ID = 'image'
```
I think `IMAGE_ID` should be only one.
However, I have no idea to make it.
I'll recommend to add comment at that time.

File: components/sendImageForm.tsx [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  const [fileName, setFileName] = useState('')
+  const [imageSource, setImageSource] = useState('')
+
+  const selectFile = () => {
+    if (!fileInputRef.current) {
+      return
+    }
+    // ローカルフォルダから画像選択ダイアログを表示する。
+    fileInputRef.current.click()
+  }
```
Comment (Marukome0743) [inline components/sendImageForm.tsx:23]: AYK, this value is not used.
Why do you leave this codes?

File: components/sendImageForm.tsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    setFileName(file.name)
+    fileReader.onload = () => {
+      setImageSource(fileReader.result as string)
+    }
+    fileReader.readAsDataURL(file)
+  }
+
+  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const files = e.target.files
+    if (!files || files.length <= 0) return
```
Comment (Marukome0743) [inline components/sendImageForm.tsx:38]:
```suggestion
    if (!files.length) return
```
I think the above if statement also meets the condition.
Are there any reasons to use double check?

File: components/uploadFormHook.tsx [hunk 6]
```diff
@@ -89,20 +88,9 @@ export const UploadFormHook: FC = () => {
...
-            capture="environment"
-            {...register('photo')}
-            required={true}
-          />
-          <label />
+          <div>
+            <SendImageForm />
+          </div>
         </div>
         <button
```
Comment (Marukome0743) [inline components/uploadFormHook.tsx:95]:
```suggestion
        </div>
        <br />
        <button
```
IMO the buttons would be separated with some space.

**Review by Marukome0743 — CHANGES_REQUESTED** [review]
Thank you for your great effort💪
I'm glad to check my review😀

File: components/sendImageForm.tsx [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+const IMAGE_ID = 'image'
+
+export const SendImageForm: React.FC = () => {
+  const fileInputRef = useRef<HTMLInputElement>(null)
+
+  // 添付画像を状態管理
+  const [imageFile, setImageFile] = useState<File | null>(null)
+  const [fileName, setFileName] = useState('')
+  const [imageSource, setImageSource] = useState('')
```
Comment (Marukome0743) [inline components/sendImageForm.tsx:15]:
```suggestion
  // 添付画像を状態管理
  const [images, setImages] = useState<{
    file: File | null
    name: string
    source: string
  }>({
    file: null,
    name: '',
    source: '',
  })

  // ファイルが読み込まれた際に、画像データを抽出する処理
  const generateImageSource = (files: FileList) => {
    const file = files[0]
    const fileReader = new FileReader()
    fileReader.onload = () => {
      setImages({
        ...images,
        file: file,
        name: file.name,
        source: fileReader.result as string,
      })
    }
    fileReader.readAsDataURL(file)
  }

  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const files = e.target.files
    if (!files || files.length <= 0) {
      return
    }
    // 画像データを抽出する処理
    generateImageSource(files)
  }
```
Refactored

**Review by Marukome0743 — CHANGES_REQUESTED** [review]
I refactored multi `useState`.
What do you think🧐

File: components/sendImageForm.tsx (hunk 4 above)
Comment (kazutan1230) [reply components/sendImageForm.tsx]: I leave it because someday the value may be used.

Comment (Marukome0743) [reply components/sendImageForm.tsx]: Do not leave unused codes without comments.
Add any comments if you do that, or leave at your local repo.

File: components/imageInput.tsx (hunk 1 above)
Comment (kazutan1230) [reply components/imageInput.tsx]: ok.

File: components/imageInput.tsx (hunk 2 above)
Comment (kazutan1230) [reply components/imageInput.tsx]: I changed using in imagePreview.tsx and uploadFormHook.tsx.

File: components/sendImageForm.tsx (hunk 3 above)
Comment (kazutan1230) [reply components/sendImageForm.tsx]: ok.

File: components/sendImageForm.tsx (hunk 7 above)
Comment (kazutan1230) [reply components/sendImageForm.tsx]: I think your code is ok and commited it.

File: components/sendImageForm.tsx (hunk 4 above)
Comment (kazutan1230) [reply components/sendImageForm.tsx]: ok, I added comments.

File: components/uploadFormHook.tsx (hunk 6 above)
Comment (kazutan1230) [reply components/uploadFormHook.tsx]: I commited this suggestion to the code.

**Comment by Marukome0743** [comment]
Could you resolve the conflicts to merge?

**Comment by kazutan1230** [comment]
> Could you resolve the conflicts to merge?

Yes, I did it.

File: components/imagePreview.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+  const handleFileChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const files = e.target.files
+    if (!files || files.length <= 0) {
+      return
+    }
+
+    // 画像データを抽出する処理
+    generateImageSource(files)
+  }
```
Comment (Marukome0743) [inline components/imagePreview.tsx:53]: `// ファイルが読み込まれた際に、画像データを抽出する処理`
`// 画像データを抽出する処理`
These comments are duplicated and redundant.
Remove a comment or combine into a function without a helper function.

File: components/imagePreview.tsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+      <div>
+        {/* hidden属性付けて、画像アップロードされてる時だけ表示できるようにした方がいい */}
+        <button
+          type="button"
+          onClick={handleClickCancelButton}
+          className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded-full mr-2"
+        >
+          × 画像アップロードキャンセル
+        </button>
+      </div>
```
Comment (Marukome0743) [inline components/imagePreview.tsx:105]: Would you fix the issues which I mentioned 2 days ago?
e.g. hidden, color(change blue to red), size...

**Review by Marukome0743 — CHANGES_REQUESTED** [review]
I appreciate your update.
Could you fix some issues which I told you 2 days ago?

File: components/imagePreview.tsx (hunk 8 above)
Comment (kazutan1230) [reply components/imagePreview.tsx]: I did.

File: components/imagePreview.tsx (hunk 9 above)
Comment (kazutan1230) [reply components/imagePreview.tsx]: I did.

File: components/imageInput.tsx [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import type React from 'react'
+import { type ComponentPropsWithRef, forwardRef } from 'react'
+import type { FC, InputHTMLAttributes } from 'react'
+
+export type Props = ComponentPropsWithRef<'input'> & {
```
Comment (Marukome0743) [inline components/imageInput.tsx:5]: This export property is not used anywhere now.
I think `export` should be removed.

File: components/uploadFormHook.tsx [hunk 11]
```diff
@@ -2,22 +2,34 @@
 
 import type { FC } from 'react'
 import { type SubmitHandler, useForm } from 'react-hook-form'
+import { ImageInput } from './imageInput'
+import { ImagePreview } from './imagePreview'
 
 // フォームの各要素と型
-type FormData = {
+export type FormData = {
```
Comment (Marukome0743) [inline components/uploadFormHook.tsx:9]: This export property is not used anywhere now.
I think `export` should be removed.

**Review by Marukome0743 — CHANGES_REQUESTED** [review]

File: components/imageInput.tsx (hunk 10 above)
Comment (kazutan1230) [reply components/imageInput.tsx]: I removed `export`.

File: components/uploadFormHook.tsx (hunk 11 above)
Comment (kazutan1230) [reply components/uploadFormHook.tsx]: I removed `export`.

**Review by Marukome0743 — APPROVED** [review]
NarutoThumbsupGIF

**Pull request merged**